src/Rules/Documentation.hs

```diff
@@ -36,12 +119,72 @@ buildPackageDocumentation context@Context {..} = when (stage == Stage1) $ do
         let haddockWay = if dynamicPrograms then dynamic else vanilla
         build $ target (context {way = haddockWay}) Haddock srcs [file]
 
-    when (package == haddock) $ haddockHtmlLib %> \_ -> do
-        let dir = takeDirectory haddockHtmlLib
-        liftIO $ removeFiles dir ["//*"]
-        copyDirectory "utils/haddock/haddock-api/resources/html" dir
+-- | Build the haddocks for GHC's libraries
+buildLibraryDocumentation :: Rules ()
+buildLibraryDocumentation = do
+    "//" ++ htmlRoot -/- "libraries/index.html" %> \file -> do
+        haddocks <- allHaddocks
+        need haddocks
+        let libDocs = filter (\x -> takeFileName x /= "ghc.haddock") haddocks
+        let iface haddock = "--read-interface="
+                            ++ (takeFileName . takeDirectory) haddock ++ ","
+                            ++ haddock
+        runBuilder Haddock ([ "--gen-index"
```

Comment:
Looks like this calls for something like `HaddockMode = BuildPackage | BuildIndex`?